src/query/storages/fuse/src/operations/common/processors/multi_table_insert_commit.rs

```diff
@@ -0,0 +1,171 @@
+        tables: HashMap<u64, Arc<dyn Table>>,
+        ctx: Arc<dyn TableContext>,
+        overwrite: bool,
+        update_stream_meta: Vec<UpdateStreamMetaReq>,
+        deduplicated_label: Option<String>,
+        catalog: Arc<dyn Catalog>,
+    ) -> Self {
+        Self {
+            commit_metas: Default::default(),
+            tables,
+            ctx,
+            overwrite,
+            update_stream_meta,
+            deduplicated_label,
+            catalog,
+        }
+    }
+}
+
+#[async_trait]
+impl AsyncSink for CommitMultiTableInsert {
+    const NAME: &'static str = "CommitMultiTableInsert";
+
+    #[async_backtrace::framed]
+    async fn on_finish(&mut self) -> Result<()> {
+        let mut update_table_meta_reqs = Vec::with_capacity(self.commit_metas.len());
+        let mut table_infos = Vec::with_capacity(self.commit_metas.len());
+        for (table_id, commit_meta) in std::mem::take(&mut self.commit_metas).into_iter() {
+            // generate snapshot
+            let mut snapshot_generator = AppendGenerator::new(self.ctx.clone(), self.overwrite);
+            snapshot_generator.set_conflict_resolve_context(commit_meta.conflict_resolve_context);
+            let table = self.tables.get(&table_id).unwrap();
+            let fuse_table = FuseTable::try_from_table(table.as_ref())?;
+            let previous = fuse_table.read_table_snapshot().await?;
+            let snapshot = snapshot_generator.generate_new_snapshot(
+                table.schema().as_ref().clone(),
+                fuse_table.cluster_key_meta.clone(),
+                previous,
+                Some(fuse_table.table_info.ident.seq),
+            )?;
+
+            // write snapshot
+            let dal = fuse_table.get_operator();
+            let location_generator = &fuse_table.meta_location_generator;
+            let location = location_generator
+                .snapshot_location_from_uuid(&snapshot.snapshot_id, TableSnapshot::VERSION)?;
+            dal.write(&location, snapshot.to_bytes()?).await?;
+
+            // build new table meta
+            let new_table_meta =
+                FuseTable::build_new_table_meta(&fuse_table.table_info.meta, &location, &snapshot)?;
+            let table_id = fuse_table.table_info.ident.table_id;
+            let table_version = fuse_table.table_info.ident.seq;
+
+            let req = UpdateTableMetaReq {
+                table_id,
+                seq: MatchSeq::Exact(table_version),
+                new_table_meta,
+                copied_files: None,
+                deduplicated_label: None,
+                update_stream_meta: vec![],
+            };
+            update_table_meta_reqs.push(req);
+            table_infos.push(table.get_table_info());
+        }
+        let is_active = self.ctx.txn_mgr().lock().is_active();
+        match is_active {
+            true => {
+                update_table_meta_reqs[0].update_stream_meta =
+                    std::mem::take(&mut self.update_stream_meta);
```

Comment:
is it possible for `on_finish` to be triggered without any commit_meta being collected? in such cases, `update_table_meta_reqs` might be of zero length.

It would be better if there are comments explaining why choosing `update_table_meta_reqs[0].update_stream_meta` as the target to update / carry the `update_stream_meta`.  e.g. "any one of the reqs may carry the update_stream_meta,  we arbitrarily choose the first one... ".

Reply:
Thanks. Some comments are added.